dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/DynatraceMetricExporter.java

```diff
@@ -102,30 +126,98 @@ public CompletableResultCode export(@Nonnull Collection<MetricData> metrics) {
+      switch (metric.getType()) {
+        case LONG_GAUGE:
+          metricLines.addAll(serializer.createLongGaugeLines(metric));
+          break;
+        case LONG_SUM:
+          isDelta = isDeltaTemporality(metric.getLongSumData().getAggregationTemporality());
+          metricLines.addAll(serializer.createLongSumLines(metric, isDelta));
+          break;
+        case DOUBLE_GAUGE:
+          metricLines.addAll(serializer.createDoubleGaugeLines(metric));
+          break;
+        case DOUBLE_SUM:
+          isDelta = isDeltaTemporality(metric.getDoubleSumData().getAggregationTemporality());
+          metricLines.addAll(serializer.createDoubleSumLines(metric, isDelta));
+          break;
+        case SUMMARY:
+          metricLines.addAll(serializer.createDoubleSummaryLines(metric));
+          break;
+        case HISTOGRAM:
+          metricLines.addAll(serializer.createDoubleHistogramLines(metric));
+          break;
+      }
+    }
+
+    return metricLines;
+  }
+
   @VisibleForTesting
   protected CompletableResultCode export(
       Collection<MetricData> metrics, HttpURLConnection connection) {
-    String mintMetricsMessage = MetricAdapter.toMint(metrics).serialize();
-    logger.log(Level.FINEST, "Exporting: {0}", mintMetricsMessage);
+
+    List<String> metricLines = makeMetricLines(metrics);
+    String mintMetricsMessage = Joiner.on('\n').join(metricLines);
+    if (logger.isLoggable(Level.FINER)) {
+      logger.finer(String.format("Exporting metrics:\n%s", mintMetricsMessage));
+    }
     try {
       connection.setRequestMethod("POST");
       connection.setRequestProperty("Accept", "*/*; q=0");
-      connection.setRequestProperty("Authorization", "Api-Token " + apiToken);
+      if (this.apiToken != null) {
+        connection.setRequestProperty("Authorization", "Api-Token " + apiToken);
+      }
       connection.setRequestProperty("Content-Type", "text/plain; charset=utf-8");
       connection.setDoOutput(true);
       try (final OutputStream outputStream = connection.getOutputStream()) {
         outputStream.write(mintMetricsMessage.getBytes(StandardCharsets.UTF_8));
       }
       int code = connection.getResponseCode();
-      if (code != 202) {
-        logger.log(Level.WARNING, "Received error code {0} from server", code);
+      if (code < 400) {
+        String response =
+            CharStreams.toString(
+                new InputStreamReader(connection.getInputStream(), Charsets.UTF_8));
+        return handleSuccess(code, metricLines.size(), response);
+      } else {
         return CompletableResultCode.ofFailure();
       }
     } catch (Exception e) {
       logger.log(Level.WARNING, "Error while exporting", e);
       return CompletableResultCode.ofFailure();
     }
-    return CompletableResultCode.ofSuccess();
+  }
+
+  private CompletableResultCode handleSuccess(int code, int totalLines, String response) {
+    if (code == 202) {
+      if (IS_NULL_ERROR_RESPONSE.matcher(response).find()) {
```

Comment:
Can you add a comment what "null error response" signifies? When i first read it I read it as "an error caused by some null" and not "there is no error"

Reply:
Renamed it to `PATTERN_RETURNED_ERROR_IS_NULL` to clarify the intention.

Reply:
I think i would go even more specific like `RETURNED_ERROR_FIELD_IS_NULL`.  "Returned error" still sounds to me like it implies an error was returned.